### ISSUE TITLE
Out of Memory Error for CRDNN architecture

### ISSUE THREAD

**Issue by KacperKubara** [speechbrain/speechbrain#1506]
Hi!

For my research project, I am finetuning a CRDNN model (from Speechbrain) to see if certain subset of audio samples will affect the model's performance in a different way. 

This means that I will have .fit() function triggered few times in a training loop. After ~two train() calls, I am getting OOM error from the model, however.

To debug it, I checked what tensors are 'hanging' after each loop iteration using this approach: https://discuss.pytorch.org/t/how-to-debug-causes-of-gpu-memory-leaks/6741

When I compare first and the second call of the brain.fit() method, additional tensors with their count are:

``` json
{
    "<class 'torch.Tensor'> torch.Size([128, 128, 3, 3])": 2,
    "<class 'torch.Tensor'> torch.Size([128])": 4,
    "<class 'torch.Tensor'> torch.Size([4096, 1024])": 16,
    "<class 'torch.Tensor'> torch.Size([512, 512])": 2,
    "<class 'torch.Tensor'> torch.Size([4096])": 32,
    "<class 'torch.Tensor'> torch.Size([1024, 2048])": 2,
    "<class 'torch.Tensor'> torch.Size([512])": 12,
    "<class 'torch.Tensor'> torch.Size([1000, 1024])": 2,
    "<class 'torch.Tensor'> torch.Size([1000])": 4,
    "<class 'torch.Tensor'> torch.Size([160, 193])": 2,
    "<class 'torch.Tensor'> torch.Size([1024, 1024])": 2,
    "<class 'torch.Tensor'> torch.Size([1000, 128])": 2,
    "<class 'torch.Tensor'> torch.Size([40, 128])": 8,
    "<class 'torch.Tensor'> torch.Size([3072, 1152])": 2,
    "<class 'torch.Tensor'> torch.Size([256, 256, 3, 3])": 2,
    "<class 'torch.Tensor'> torch.Size([4096, 2048])": 12,
    "<class 'torch.Tensor'> torch.Size([3, 182])": -2,
    "<class 'torch.Tensor'> torch.Size([1024, 10])": 2,
    "<class 'torch.Tensor'> torch.Size([128, 1, 3, 3])": 2,
    "<class 'torch.Tensor'> torch.Size([1, 1024])": 2,
    "<class 'torch.Tensor'> torch.Size([1000, 512])": 2,
    "<class 'torch.Tensor'> torch.Size([512, 2048])": 2,
    "<class 'torch.Tensor'> torch.Size([3072, 1024])": 2,
    "<class 'torch.Tensor'> torch.Size([3072])": 4,
    "<class 'torch.Tensor'> torch.Size([3, 182, 1024])": -1,
    "<class 'torch.Tensor'> torch.Size([256])": 4,
    "<class 'torch.Tensor'> torch.Size([1024, 512])": 4,
    "<class 'torch.Tensor'> torch.Size([160, 193, 1024])": 1,
    "<class 'torch.Tensor'> torch.Size([1024])": 10,
    "<class 'torch.Tensor'> torch.Size([20, 256])": 8,
    "<class 'torch.Tensor'> torch.Size([10, 1, 201])": 2,
    "<class 'torch.Tensor'> torch.Size([4096, 2560])": 4,
    "<class 'torch.Tensor'> torch.Size([256, 128, 3, 3])": 2
}
```

And the architecture parameters for the model in the YAML config file are:
``` yaml
# dynamic batching parameters, if used
dynamic_batch_sampler:
   feats_hop_size: 0.01
   max_batch_len: 20000 # in terms of frames
   shuffle_ex: True
   batch_ordering: random
   num_buckets: 20

# Feature parameters
sample_rate: 16000
n_fft: 400
n_mels: 40

opt_class: !name:torch.optim.Adadelta
   lr: !ref <lr>
   rho: 0.95
   eps: 1.e-8

# opt_class: !name:torch.optim.SGD
#    lr: 0.00001

# Dataloader options
train_dataloader_opts:
   batch_size: !ref <batch_size>

valid_dataloader_opts:
   batch_size: !ref <batch_size>

test_dataloader_opts:
   batch_size: !ref <batch_size>

# Model parameters
activation: !name:torch.nn.LeakyReLU
dropout: 0.15
cnn_blocks: 2
cnn_channels: (128, 256)
inter_layer_pooling_size: (2, 2)
cnn_kernelsize: (3, 3)
time_pooling_size: 4
rnn_class: !name:speechbrain.nnet.RNN.LSTM
rnn_layers: 4
rnn_neurons: 1024
rnn_bidirectional: True
dnn_blocks: 2
dnn_neurons: 512
emb_size: 128
dec_neurons: 1024
output_neurons: 1000  # Number of tokens (same as LM)
blank_index: 0
bos_index: 0
eos_index: 0
unk_index: 0

# Decoding parameters
min_decode_ratio: 0.0
max_decode_ratio: 1.0
valid_beam_size: 40
test_beam_size: 40
eos_threshold: 1.5
using_max_attn_shift: True
max_attn_shift: 240
lm_weight: 0.50
ctc_weight_decode: 0.0
coverage_penalty: 1.5
temperature: 1.25
temperature_lm: 1.25

epoch_counter: !new:speechbrain.utils.epoch_loop.EpochCounter
   limit: !ref <number_of_epochs>

normalize: !new:speechbrain.processing.features.InputNormalization
   norm_type: global

compute_features: !new:speechbrain.lobes.features.Fbank
   sample_rate: !ref <sample_rate>
   n_fft: !ref <n_fft>
   n_mels: !ref <n_mels>


augmentation: !new:speechbrain.lobes.augment.TimeDomainSpecAugment
   sample_rate: !ref <sample_rate>
   speeds: [95, 100, 105]

enc: !new:speechbrain.lobes.models.CRDNN.CRDNN
   input_shape: [null, null, !ref <n_mels>]
   activation: !ref <activation>
   dropout: !ref <dropout>
   cnn_blocks: !ref <cnn_blocks>
   cnn_channels: !ref <cnn_channels>
   cnn_kernelsize: !ref <cnn_kernelsize>
   inter_layer_pooling_size: !ref <inter_layer_pooling_size>
   time_pooling: True
   using_2d_pooling: False
   time_pooling_size: !ref <time_pooling_size>
   rnn_class: !ref <rnn_class>
   rnn_layers: !ref <rnn_layers>
   rnn_neurons: !ref <rnn_neurons>
   rnn_bidirectional: !ref <rnn_bidirectional>
   rnn_re_init: True
   dnn_blocks: !ref <dnn_blocks>
   dnn_neurons: !ref <dnn_neurons>
   use_rnnp: False

emb: !new:speechbrain.nnet.embedding.Embedding
   num_embeddings: !ref <output_neurons>
   embedding_dim: !ref <emb_size>

dec: !new:speechbrain.nnet.RNN.AttentionalRNNDecoder
   enc_dim: !ref <dnn_neurons>
   input_size: !ref <emb_size>
   rnn_type: gru
   attn_type: location
   hidden_size: !ref <dec_neurons>
   attn_dim: 1024
   num_layers: 1
   scaling: 1.0
   channels: 10
   kernel_size: 100
   re_init: True
   dropout: !ref <dropout>

ctc_lin: !new:speechbrain.nnet.linear.Linear
   input_size: !ref <dnn_neurons>
   n_neurons: !ref <output_neurons>

seq_lin: !new:speechbrain.nnet.linear.Linear
   input_size: !ref <dec_neurons>
   n_neurons: !ref <output_neurons>

log_softmax: !new:speechbrain.nnet.activations.Softmax
   apply_log: True

ctc_cost: !name:speechbrain.nnet.losses.ctc_loss
   blank_index: !ref <blank_index>

seq_cost: !name:speechbrain.nnet.losses.nll_loss
   label_smoothing: 0.1

# This is the RNNLM that is used according to the Huggingface repository
# NB: It has to match the pre-trained RNNLM!!
lm_model: !new:speechbrain.lobes.models.RNNLM.RNNLM
   output_neurons: !ref <output_neurons>
   embedding_dim: !ref <emb_size>
   activation: !name:torch.nn.LeakyReLU
   dropout: 0.0
   rnn_layers: 2
   rnn_neurons: 2048
   dnn_blocks: 1
   dnn_neurons: 512
   return_hidden: True  # For inference

tokenizer: !new:sentencepiece.SentencePieceProcessor

# Models
modules:
   enc: !ref <enc>
   emb: !ref <emb>
   dec: !ref <dec>
   ctc_lin: !ref <ctc_lin>
   seq_lin: !ref <seq_lin>
   normalize: !ref <normalize>
   lm_model: !ref <lm_model>

model: !new:torch.nn.ModuleList
   - [!ref <enc>, !ref <emb>, !ref <dec>, !ref <ctc_lin>, !ref <seq_lin>]

valid_search: !new:speechbrain.decoders.S2SRNNBeamSearcher
   embedding: !ref <emb>
   decoder: !ref <dec>
   linear: !ref <seq_lin>
   ctc_linear: !ref <ctc_lin>
   bos_index: !ref <bos_index>
   eos_index: !ref <eos_index>
   blank_index: !ref <blank_index>
   min_decode_ratio: !ref <min_decode_ratio>
   max_decode_ratio: !ref <max_decode_ratio>
   beam_size: !ref <valid_beam_size>
   eos_threshold: !ref <eos_threshold>
   using_max_attn_shift: !ref <using_max_attn_shift>
   max_attn_shift: !ref <max_attn_shift>
   coverage_penalty: !ref <coverage_penalty>
   temperature: !ref <temperature>

test_search: !new:speechbrain.decoders.S2SRNNBeamSearchLM
   embedding: !ref <emb>
   decoder: !ref <dec>
   linear: !ref <seq_lin>
   ctc_linear: !ref <ctc_lin>
   language_model: !ref <lm_model>
   bos_index: !ref <bos_index>
   eos_index: !ref <eos_index>
   blank_index: !ref <blank_index>
   min_decode_ratio: !ref <min_decode_ratio>
   max_decode_ratio: !ref <max_decode_ratio>
   beam_size: !ref <test_beam_size>
   eos_threshold: !ref <eos_threshold>
   using_max_attn_shift: !ref <using_max_attn_shift>
   max_attn_shift: !ref <max_attn_shift>
   coverage_penalty: !ref <coverage_penalty>
   lm_weight: !ref <lm_weight>
   ctc_weight: !ref <ctc_weight_decode>
   temperature: !ref <temperature>
   temperature_lm: !ref <temperature_lm>

lr_annealing: !new:speechbrain.nnet.schedulers.NewBobScheduler
   initial_value: !ref <lr>
   improvement_threshold: 0.0025
   annealing_factor: 0.8
   patient: 0

checkpointer: !new:speechbrain.utils.checkpoints.Checkpointer
   checkpoints_dir: !ref <save_folder>
   recoverables:
      model: !ref <model>
      scheduler: !ref <lr_annealing>
      normalizer: !ref <normalize>
      counter: !ref <epoch_counter>

train_logger: !new:speechbrain.utils.train_logger.FileTrainLogger
   save_file: !ref <train_log>

error_rate_computer: !name:speechbrain.utils.metric_stats.ErrorRateStats

cer_computer: !name:speechbrain.utils.metric_stats.ErrorRateStats
   split_tokens: True

# The pretrainer allows a mapping between pretrained files and instances that
# are declared in the yaml. E.g here, we will download the file lm.ckpt
# and it will be loaded into "lm" which is pointing to the <lm_model> defined
# before.
pretrainer: !new:speechbrain.utils.parameter_transfer.Pretrainer
   collect_in: !ref <save_folder>
   loadables:
      lm: !ref <lm_model>
      tokenizer: !ref <tokenizer>
   paths:
      lm: !ref <pretrained_lm_tokenizer_path>/lm.ckpt
      tokenizer: !ref <pretrained_lm_tokenizer_path>/tokenizer.ckpt
```

Do you have any idea why those tensor appear and (likely) cause the OOM? I've tried using methods like `gc.collect()` or `torch.cuda.empty_cache()` but they don't work.

I am quite stuck on this problem and I am not sure how to even approach it to solve this issue. Any help greatly appreciated!

**Comment by TParcollet** [comment]
Hi, I am sorry, but I am not even sure to properly understand what causes these logs :/ Could you explicit a bit more what we are witnessing here, in term of actual experiment and pipeline?

**Comment by KacperKubara** [comment]
Hmm, I actually think I have just fixed that issue. I need to test it a bit more but I don't get the OOM anymore.

The problem was in 2 areas, I think:

1) Audios were slightly too long. After changing max audio length from 10s to 8s, it seems to be okay now
2) I was saving part of the computational graph to get gradients from certain layer in the model (part of my experiments). During the training it consumed too much memory I believe, so I had to turn it off as well!
